lxd/endpoints/endpoints.go

```diff
@@ -0,0 +1,286 @@
+	servers   map[kind]*http.Server // HTTP servers by endpoint type.
+	cert      *shared.CertInfo      // Keypair and CA to use for TLS.
+
+	systemdListenFDsStart int // First socket activation FD, for tests.
+}
+
+// Up brings up all configured endpoints and starts accepting HTTP requests.
+func (e *Endpoints) up(config *Config) error {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	e.servers = map[kind]*http.Server{
+		devlxd:  config.DevLxdServer,
+		local:   config.RestServer,
+		network: config.RestServer,
+	}
+	e.cert = config.Cert
+
+	var err error
+
+	// Check for socket activation.
+	systemdListeners := util.GetListeners(e.systemdListenFDsStart)
+	if len(systemdListeners) > 0 {
+		logger.Infof("LXD is socket activated")
+		e.listeners = activatedListeners(systemdListeners, e.cert)
+	} else {
+		logger.Infof("LXD isn't socket activated")
+		e.listeners = map[kind]net.Listener{}
+
+		e.listeners[local], err = localCreateListener(config.Dir, config.LocalUnixSocketGroup)
+		if err != nil {
+			return fmt.Errorf("local endpoint: %v", err)
+		}
+	}
+
+	// Start the devlxd listener
+	e.listeners[devlxd], err = createDevLxdlListener(config.Dir)
+	if err != nil {
+		return err
+	}
+
+	if config.NetworkAddress != "" {
+		listener, ok := e.listeners[network]
+		if ok {
+			logger.Infof("Replacing inherited TCP socket with configured one")
+			listener.Close()
+		}
+		// Errors here are not fatal and are just logged.
+		e.listeners[network] = networkCreateListener(config.NetworkAddress, e.cert)
+	}
+
+	logger.Infof("Starting /dev/lxd handler:")
+	e.serveHTTP(devlxd)
+
+	logger.Infof("REST API daemon:")
+	e.serveHTTP(local)
+	e.serveHTTP(network)
+
+	return nil
+}
+
+// Down brings down all endpoints and stops serving HTTP requests.
+func (e *Endpoints) Down() error {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	logger.Infof("Stopping REST API handler:")
+	err := e.closeListener(network)
+	if err != nil {
+		return err
+	}
+	err = e.closeListener(local)
+	if err != nil {
+		return err
+	}
+
+	logger.Infof("Stopping /dev/lxd handler")
+	err = e.closeListener(devlxd)
+	if err != nil {
+		return err
+	}
+
+	if e.tomb != nil {
+		e.tomb.Kill(nil)
+		e.tomb.Wait()
+	}
+
+	return nil
+}
+
+// Start an HTTP server for the endpoint associated with the given code.
+func (e *Endpoints) serveHTTP(kind kind) {
+	listener := e.listeners[kind]
+
+	if listener == nil {
+		return
+	}
+
+	message := fmt.Sprintf(" - binding %s", descriptions[kind])
+	logger.Info(message, log.Ctx{"socket": listener.Addr()})
+
+	server := e.servers[kind]
+
+	// Defer the creation of the tomb, so Down() doesn't wait on it unless
+	// we actually have spawned at least a server.
+	if e.tomb == nil {
+		e.tomb = &tomb.Tomb{}
+	}
+
+	e.tomb.Go(func() error {
+		server.Serve(listener)
+		return nil
+	})
+}
+
+// Stop the HTTP server of the endpoint associated with the given code. The
+// associated socket will be shutdown too.
+func (e *Endpoints) closeListener(code kind) error {
```

Comment:
why is code being used here instead of kind?

Reply:
No good reason, I guess (except that it shadows the type, so it can't be used anymore in the block). Fixed in 0f6be5592107900b9a5f7bea76afd96808824576.